File: java/hashtableseparatechaining/src/main/java/com/js/datastructuresalgorithms/hashtableseparatechaining/JSEntry.java
package com.js.datastructuresalgorithms.hashtableseparatechaining;

public class JSEntry <K, V> {

    //*****************
    // Class Members **
    //*****************
    private final K mKey;
    private final V mValue;
    private final int mHash;

    //***************
    // Constructor **
    //***************
    public JSEntry(K key, V value) {
        mKey = key;
        mValue = value;
        mHash = key.hashCode();
    }

    public int hashCode() { return mHash; }

    public boolean equal(JSEntry<K, V> other) { return hashCode() == other.hashCode() && key().equals(other.key()); }

    public String toString() { return key() + " => " + value(); }

    public K key() { return mKey; }

    public V value() { return mValue; }
}

File: node/hashtablechaining/src/types/JSEntry.js
const crypto = require('crypto');

class JSEntry {

    #mKey
    #mValue
    #mHash

    constructor(key, value) {
        this.#mKey = key;
        this.#mValue = value;
        this.#mHash = null;
    }

    get key() { return this.#mKey };

    get value() { return this.#mValue; }

    get hash() { return this.#mHash === null ? this.#mHash = this.#hash(this.key) : this.#mHash; }

    equal = (that) => this.hash === that.hash && this.key === that.key;

    //*******************
    // Private Members **
    //*******************

    #hash = (value) => parseInt(crypto.createHash('md5').update(this.key).digest('hex'), 16) % 100000000;
}

module.exports = JSEntry;